__tests__/test-utils/setup-e2e.js

```diff
@@ -0,0 +1,484 @@
+function recreateEndToEndTestResultDirectory () {
+  const testFolderFilename = getTestFolderFilename()
+  fs.removeSync(testFolderFilename)
+  fs.ensureDirSync(testFolderFilename)
+}
+
+/**
+ * Verifies that the needed items in order to run the e2e tests locally have
+ * been setup properly.
+ */
+async function verifySetupForLocalEnvironment () {
+  const errors = []
+
+  // make sure e2e.yml exists
+  try {
+    await fs.stat('configurations/end-to-end/env.yml')
+  } catch (e) {
+    errors.push(new Error('Failed to detect file `configurations/end-to-end/env.yml`'))
+  }
+
+  // make sure services are running
+  const endpointChecks = [
+    {
+      name: 'Front-end server',
+      url: 'http://localhost:9966'
+    }, {
+      name: 'Back-end server',
+      url: 'http://localhost:4000'
+    }, {
+      name: 'OTP server',
+      url: 'http://localhost:8080'
+    }
+  ]
+
+  await Promise.all(
+    endpointChecks.map(
+      endpoint => (
+        // return an executed promise
+        async () => {
+          try {
+            const fetchResult = await fetch(endpoint.url)
+            if (fetchResult.status >= 400) {
+              throw new Error('Bad response')
+            }
+          } catch (e) {
+            errors.push(
+              new Error(
+                `Unable to fetch from ${endpoint.name} (${endpoint.url})`
+              )
+            )
+          }
+        }
+      )()
+    )
+  )
+
+  // TODO: verify that aws credentials have been setup properly
+
+  if (errors.length > 0) {
+    throw new AggregateError(errors)
+  }
+}
+
+/**
+ * Sets up the needed items in order to be ran in a continuous integration
+ * environment. For now, Travis CI is the only CI environment that this code is
+ * built for.
+ */
+function setupForContinuousIntegrationEnvironment () {
+  return Promise.all([
+    startBackendServer(),
+    startClientServer(),
+    startOtp()
+  ])
+}
+
+/**
+ * Start a server to collect coverage from the e2e tests.
+ */
+function startCoverageServer () {
+  return new Promise((resolve, reject) => {
+    console.log('starting coverage server')
+
+    try {
+      spawnDetachedProcess(
+        'node',
+        ['__tests__/test-utils/e2e-coverage-collector-server.js'],
+        'e2e-coverage-collector-server'
+      )
+    } catch (e) {
+      console.error(`error starting coverage server: ${e}`)
+      return reject(e)
+    }
+
+    console.log('successfully started coverage server')
+    resolve()
+  })
+}
+
+module.exports = async function () {
+  // syncrhonosuly recreate the test results directory in to avoid race
+  // conditions
+  recreateEndToEndTestResultDirectory()
+
+  // do different setup depending on runtime environment
+  const setupItems = []
+  if (isCi) {
+    setupItems.push(setupForContinuousIntegrationEnvironment())
+  } else {
+    setupItems.push(verifySetupForLocalEnvironment())
+  }
+  if (collectingCoverage) setupItems.push(startCoverageServer())
+
+  return Promise.all(setupItems)
+}
```

Comment:
Does the module.exports need to be defined at the end of the file? If not, I kind of like the idea of moving it to the top and giving it a comment about when/where it is called.

Reply:
In certain cases it'd be fine to have it at the top, but I think in general it's a good idea to have it at the bottom because not everything gets hoisted. See https://stackoverflow.com/questions/23791749/why-must-a-modules-exports-be-declared-at-the-bottom-of-a-file and https://scotch.io/tutorials/understanding-hoisting-in-javascript. I'll add comments about this exported function tho.